metricq_wizard_backend/metricq/cluster_scanner.py

```diff
@@ -0,0 +1,521 @@
+    async def _run_scan(self) -> None:
+        # TODO add check for (db) queue bindings
+        # TODO add check for token names
+
+        assert self.db_metadata is not None
+
+        async with HistoryClient(self.token, self.url, add_uuid=True) as client:
+            tasks = AsyncTaskPool(max_tasks=250)
+
+            async for doc in self.db_metadata.docs():
+                metric = doc.id
+                metadata = doc.data
+
+                # this assert is purely for mypy. doc.data can only return None
+                # is the doc does not exist. Which it clearly does, as we only
+                # get existing documents from the docs iterator.
+                assert metadata is not None
+
+                await tasks.append(self.check_metric_metadata(metric, metadata))
+
+                if metadata.get("historic", False):
+                    # Only check the db status for historic metrics
+                    await tasks.append(
+                        self.check_metric_is_dead(client, metric, metadata)
+                    )
+                    await tasks.append(
+                        self.check_metric_for_infinites(client, metric, metadata)
+                    )
+
+                # there is no tooling for renaming metrics, so bad
+                # names is nothing we should warn about yet.
+                # await tasks.append(self.check_metric_name(metric))
+
+            await tasks.completed()
+
+    async def create_issue_report(
+        self,
+        type: str,
+        scope_type: str,
+        scope: str,
+        date: str | None = None,
+        severity: str | None = None,
+        **kwargs: Any,
+    ):
+        assert self.db_issues is not None
+        report = await self.db_issues.create(
+            id=f"{type}-{scope_type}-{scope}", exists_ok=True
+        )
+
+        report["severity"] = "warning" if severity is None else severity
+
+        if "first_detection_date" not in report:
+            # only set first_detection_date when creating the report, not
+            # on updates
+            report["first_detection_date"] = (
+                str(Timestamp.now().datetime.isoformat()
+                    ) if date is None else date
+            )
+
+        report["date"] = (
+            str(Timestamp.now().datetime.isoformat()) if date is None else date
+        )
+
+        report["type"] = type
+        report["scope_type"] = scope_type
+        report["scope"] = scope
+
+        # entries in the new kwargs should overwrite existing entries
+        for key, val in kwargs.items():
+            report[key] = val
+
+        await report.save()
+
+    async def delete_issue_report(
+        self,
+        type: str,
+        scope_type: str,
+        scope: str,
+    ):
+        assert self.db_issues is not None
+
+        report = await self.db_issues.create(
+            id=f"{type}-{scope_type}-{scope}", exists_ok=True
+        )
+        if report.exists:
+            await report.delete()
```

Comment:
So it's not an issue at all to delete what not exists?

Reply:
(Why) do we not need to raise an error if one tries to delete a report that does not exist?

Reply:
Yes. As the checks iterate over metrics. In a check I don't know if an error report existed, which has been fixed. Hence, I'm delete regardless whether one existed. Raising an error here would simply make me go around and sprinkle `with suppress` everywhere.